### PR TITLE
Public types for Duration and Point

### PR DESCRIPTION
**Pull request by sherfert**

### CODE REVIEW

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/temporal/Duration.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.List;
+
+import static java.time.temporal.ChronoUnit.DAYS;
+import static java.time.temporal.ChronoUnit.MONTHS;
+import static java.time.temporal.ChronoUnit.NANOS;
+import static java.time.temporal.ChronoUnit.SECONDS;
+import static java.util.Arrays.asList;
+import static java.util.Collections.unmodifiableList;
+
+public interface Duration extends TemporalAmount
```
Comment (thobe) [inline community/graphdb-api/src/main/java/org/neo4j/graphdb/temporal/Duration.java:36]: Why is this needed? What value does this provide to the user that they do not already get from `TemporalAmount`?

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/CRS.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.graphdb.spatial.impl;
```
Comment (thobe) [inline community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/CRS.java:20]: This is a new pattern to `impl` packages - if this is implementation, why is it part of the API?

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/Point.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.List;
+import java.util.Objects;
+
+import org.neo4j.graphdb.spatial.CRS;
+import org.neo4j.graphdb.spatial.Coordinate;
+
+/**
+ * Default implementation of a {@link org.neo4j.graphdb.spatial.Point}
+ */
+public final class Point implements org.neo4j.graphdb.spatial.Point
```
Comment (thobe) [inline community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/Point.java:32]: Why does the API definition module contain a "default implementation" - who is it for?

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/temporal/impl/Duration.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.graphdb.temporal.impl;
+
+public class Duration implements org.neo4j.graphdb.temporal.Duration
```
Comment (thobe) [inline community/graphdb-api/src/main/java/org/neo4j/graphdb/temporal/impl/Duration.java:22]: This means that we now have *two* implementations of `org.neo4j.graphdb.temporal.Duration`. This seems strange to me, an interface that there is probably no need for, with two separate implementations. Who is helped by this?

**Review by thobe — CHANGES_REQUESTED** [review]
I would like to strongly question the need for this.

I can see the value in having a concrete type for `Point`, but why complicate it with both an interface and an implementation class? The other properties in Neo4j aren't defined that way, `java.lang.String` doesn't have an interface that one codes to.

For `Duration` I don't see the point at all, `TemporalAmount` is perfectly adequate for the purpose already.

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/temporal/Duration.java (hunk 1 above)
Comment (sherfert) [reply community/graphdb-api/src/main/java/org/neo4j/graphdb/temporal/Duration.java:36]: I created this interface to share the code between DurationValue and `graphdb.temporal.impl.Duration`. Other than that, it's not needed.

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/CRS.java (hunk 2 above)
Comment (sherfert) [reply community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/CRS.java:20]: I don't care about the package name, but I want to provide implementations to users of Core API and procedure writers. And these implementations should not be from the Values module.

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/Point.java (hunk 3 above)
Comment (sherfert) [reply community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/Point.java:32]: See above. Also, all downstream repositories like docs, jmh-benchmarks,... which constantly break because of using internal types instead of public API. So let's change that

**Comment by sherfert** [comment]
`TemporalAmount` has two implementations from Java's side, but none of these two suffices to capture everything that can be represented using a `DurationValue`. WithCore API or procedures or whenever you want to pass in a Duration as a parameter in Cypher, you need an implementation of TemporalAmount that is rich enough for all use cases. And one that is public and not suddenly subject to change, which can happen to `DurationValue`.

The reason for the Point interface is simply that this one existed before. It's not needed, but we can't remove it in 3.4, we could at most deprecate it. The Duration interface is for sharing code.

File: community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/Point.java (hunk 3 above)
Comment (craigtaverner) [reply community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/Point.java:32]: I previously added Point implementations to the documentation module so that it would not use PointValue.

Comment (sherfert) [reply community/graphdb-api/src/main/java/org/neo4j/graphdb/spatial/impl/Point.java:32]: I prefer all downstream modules used the same instead of coding their own.

**Comment by thobe** [comment]
I believe `TemporalAmount` itself is rich enough for all use cases. When is it not?

**Comment by sherfert** [comment]
The interface `TemporalAmount` is rich enough. But the two implementations Java provides are not (Period and Duration)

**Comment by lutovich** [comment]
I think it is better to not name our own duration interface just `Duration`. Name clashes with `java.time.Duration` and represents quite a different concept. Those users who use Cypher duration will most likely be using classes from `java.time` package and will be forced to use fully qualified class name. It might also be common to convert from Cypher duration to `java.time.Duration`. Cypher duration is a combination of `java.time.Period` and `java.time.Duration`. It also matches the ISO 8601 duration definition https://en.wikipedia.org/wiki/ISO_8601#Durations. Thus in the java driver public interface is called `IsoDuration`.

**Comment by sherfert** [comment]
This will be deferred until we have a proper design.
